### PR TITLE
refactor: Extract upper-bound binary search from `findNearestLogEventByTimestamp` into a function (resolves #343).

### LINKED ISSUE #343: Refactor: Extract binary search logic to generic upper_bound_binary_search utility
## Description

Currently, the `findNearestLogEventByTimestamp` method in `JsonlDecoder` contains binary search logic that could be extracted into a generic, reusable utility function. This would improve code reusability and make the logic unit testable.

## Suggested Implementation

Create a generic `upper_bound_binary_search` function in `src/utils/binary_search.ts` similar to [`std::upper_bound`](https://github.com/y-scope/clp-ffi-js/blob/d85aeb7801dba550a2623751e26e574d529c5bb8/src/clp_ffi_js/ir/StreamReader.hpp#L324C32-L324C43).

### Proposed Function Signature
```typescript
const upper_bound_binary_search = <T>(get: (index: number)=>T, begin: number, end: number, upperboundValue: T) => {
    ...
}
```

## Benefits
- Makes binary search logic reusable across the codebase
- Enables unit testing of the binary search implementation
- Improves code maintainability and reduces duplication

## References
- PR: https://github.com/y-scope/yscope-log-viewer/pull/152
- Comment: https://github.com/y-scope/yscope-log-viewer/pull/152#discussion_r2198244539
- Requested by: @hoophalab

### PR DESCRIPTION
**Pull request by Henry8192**
# Description

This pull request refactors the `findNearestLogEventByTimestamp` method in the `JsonlDecoder` class by introducing a reusable utility function, `upperBoundBinarySearch`, to simplify and optimize the binary search logic. Additionally, it includes unit tests for the new utility function. Below is a summary of the most important changes:

### Refactoring and Optimization:

* The `findNearestLogEventByTimestamp` core binary search logic in `JsonlDecoder` was refactored to use the newly introduced `upperBoundBinarySearch` utility function, replacing the inline binary search implementation. This improves code readability and reusability.

### Testing:

* Added comprehensive unit tests for `upperBoundBinarySearch` in `test/utils/math.test.ts`, covering edge cases such as empty arrays, single-element arrays, and arrays with duplicate values.

* [x] The PR satisfies the [contribution guidelines][yscope-contrib-guidelines].
* [x] This is a breaking change and that has been indicated in the PR title, OR this isn't a
  breaking change.
* [x] Necessary docs have been updated, OR no docs need to be updated.

# Validation performed
* Passed all tests with command `npx jest test/utils/math.test.ts`
* Tried this timestamp query on log viewer: http://localhost:3010/?filePath=https://yscope.s3.us-east-2.amazonaws.com/sample-logs/yarn-ubuntu-resourcemanager-ip-172-31-17-135.log.1.clp.zst#timestamp=1427103813789 and it jumps to the correct log event number with timestamp Mon Mar 23 2015 09:43:33.789 GMT.

[yscope-contrib-guidelines]: https://docs.yscope.com/dev-guide/contrib-guides-overview.html

## Summary by CodeRabbit

* **Refactor**
  * Nearest-log-event lookup now uses a shared upper-bound binary-search utility for clearer, more maintainable behaviour.

* **New Features**
  * Added a reusable upper-bound binary-search utility for efficient lookups on sorted collections.

* **Tests**
  * Added comprehensive tests for the new utility covering empty inputs, edge cases, boundaries and duplicates.

* **Documentation**
  * Clarified that the log-event lookup may return null when no events exist.

### CODE REVIEW

File: src/utils/math.ts [hunk 1]
```diff
@@ -20,7 +23,48 @@ const clamp = (num: number, min: number, max: number) => Math.min(Math.max(num,
...
+ * @return
+ */
+const upperBoundBinarySearch = <T>(
+    get: (index: number) => T,
+    lowIdx: number,
+    highIdx: number,
+    upperboundValue: T,
+): Nullable<number> => {
+    if (highIdx < lowIdx || "undefined" === typeof (get(highIdx))) {
+        return null;
```
Comment (quinntaylormitchell) [inline src/utils/math.ts:44]: The `null` return for empty array or invalid indices should be mentioned in the docstring.

**Review by quinntaylormitchell — CHANGES_REQUESTED** [review]
PR is looking good Henry; a few high-level comments for you as the code itself looks pretty good:
1.	I would flesh out your “Description” section. I would A) communicate the purpose for the PR, B) say whether or not you changed any logic (or if it’s just a pure refactor), and C) provide a link to issue #343.
2.	In the “Validation performed” section, the link you provided does go directly to a log result, but I’m not sure how that link by itself provides validation for the PR. On the page I see when I follow the link, I don’t see the “query” you’re referring to. I ran the query `2015-03-24T01:09:` myself and it successfully returned all 37 of the logs on page 38 that occurred during 1:09am, so maybe you can put that in the validation section (with the link) if you’d like to flesh it out.
3.	I agree with the rabbit re. improving the detail in the docstring for the function, but parameters like `lowIdx` and `highIdx` are self-explanatory and don't need explanation.
4.	I kind of agree with the rabbit re. avoiding parameter reassignment, but I also think it’s good that you stuck to the conventions about naming 0-based index variables with the `Idx` suffix. @kirkrodrigues or @junhaoliao what do you think?

File: src/utils/math.ts [hunk 2]
```diff
@@ -20,7 +23,49 @@ const clamp = (num: number, min: number, max: number) => Math.min(Math.max(num,
...
+ * are greater, or `null` if the collection is empty or the indices are invalid.
+ */
+const upperBoundBinarySearch = <T>(
+    get: (index: number) => T,
+    lowIdx: number,
+    highIdx: number,
+    upperboundValue: T,
+): Nullable<number> => {
+    if (highIdx < lowIdx || "undefined" === typeof (get(highIdx))) {
+        return null;
```
[the remark this answers is not in the document]
Comment (Henry8192) [reply src/utils/math.ts:45]: @coderabbitai highIdx and lowIdx are numbers, and it cannot be "undefined".

File: test/utils/math.test.ts [hunk 3]
```diff
@@ -22,7 +24,7 @@ describe("clamp", () => {
     });
 
     test(
-        "returns the upper boundary when the upper boundary is less than the lower boundary",
+        "returns the upper boundary if the upper boundary is less than the lower boundary",
```
Comment (quinntaylormitchell) [inline test/utils/math.test.ts:27]: Even though you're right about this change, this test case is outside of the scope of this PR, so I think you should save it for another time.

**Review by quinntaylormitchell — COMMENTED** [review]
Looks good Henry - left a comment on a change that I think is outside the scope of the PR, but other than that I think you're good to go.

File: src/utils/math.ts [hunk 4]
```diff
@@ -20,7 +23,49 @@ const clamp = (num: number, min: number, max: number) => Math.min(Math.max(num,
 const getChunkNum =
     (itemNum: number, chunkSize: number) => Math.max(1, Math.ceil(itemNum / chunkSize));
 
+/**
+ * Finds the index of the last element in a sorted collection that is less than or equal to
+ * `upperboundValue`. If all elements in the collection are greater than `upperboundValue`, return
+ * the first index of the collection (i.e., `0`).
```
Comment (kirkrodrigues) [inline src/utils/math.ts:29]:
```suggestion
 * Finds the last element in a sorted collection that is less than or equal to `upperboundValue`. If
 * all elements in the collection are greater than `upperboundValue`, returns the first index of the
 * collection (i.e., `0`).
```

File: src/utils/math.ts [hunk 5]
```diff
@@ -20,7 +23,49 @@ const clamp = (num: number, min: number, max: number) => Math.min(Math.max(num,
...
+ *
+ * @param get Function provided to access elements in the collection by index.
```
Comment (kirkrodrigues) [inline src/utils/math.ts:31]:
```suggestion
 * @param get Function to access elements in the collection by index.
```

File: src/utils/math.ts [hunk 6]
```diff
@@ -20,7 +23,49 @@ const clamp = (num: number, min: number, max: number) => Math.min(Math.max(num,
...
+ * @param lowIdx
+ * @param highIdx
```
Comment (kirkrodrigues) [inline src/utils/math.ts:33]: Since we only use this function in one place and we pass in the range of the entire array, do we really need to support passing in these indices?

File: src/utils/math.ts [hunk 7]
```diff
@@ -20,7 +23,49 @@ const clamp = (num: number, min: number, max: number) => Math.min(Math.max(num,
...
+ * @param upperboundValue
```
Comment (kirkrodrigues) [inline src/utils/math.ts:34]: How about `upperBoundValue`?

File: src/utils/math.ts [hunk 8]
```diff
@@ -20,7 +23,49 @@ const clamp = (num: number, min: number, max: number) => Math.min(Math.max(num,
...
+ * @return The index of the last element less than or equal to `upperboundValue`,`0` if all elements
+ * are greater, or `null` if the collection is empty or the indices are invalid.
```
Comment (kirkrodrigues) [inline src/utils/math.ts:36]:
```suggestion
 * @return The index of the last element less than or equal to `upperboundValue`, `0` if all
 * elements are greater, or `null` if the collection is empty or the indices are invalid.
```

File: src/utils/math.ts [hunk 9]
```diff
@@ -20,7 +23,49 @@ const clamp = (num: number, min: number, max: number) => Math.min(Math.max(num,
...
+ */
+const upperBoundBinarySearch = <T>(
+    get: (index: number) => T,
+    lowIdx: number,
+    highIdx: number,
+    upperboundValue: T,
+): Nullable<number> => {
+    if (highIdx < lowIdx || "undefined" === typeof (get(highIdx))) {
```
Comment (kirkrodrigues) [inline src/utils/math.ts:44]: Based on this condition, I guess `highIdx` is inclusive. This is the kind of detail that should be included in the docstring; otherwise users have to read the code to figure it out, negating the point of the docstring.

Comment (kirkrodrigues) [inline src/utils/math.ts:44]: If we pass in the entire array, we can just check if `length` is 0 rather than doing this "undefined" check.

**Review by kirkrodrigues — CHANGES_REQUESTED** [review]

File: src/utils/math.ts (hunk 6 above)
Comment (Henry8192) [reply src/utils/math.ts:33]: It is a question of whether we should make this function as a generic function or just the timestamp helper. Adding these indexes are intended to make this function more flexible, but yeah currently I cannot think of an example of searching partially in the array. I'll let you to decide.

Comment (kirkrodrigues) [reply src/utils/math.ts:33]: I prefer to keep it simple for now. We can always add the functionality to search only a subarray later if we need it.

**Review by kirkrodrigues — APPROVED** [review]
For the PR title, how about:

```
refactor: Extract upper-bound binary search from `findNearestLogEventByTimestamp` into a function (resolves #343).
```
